Test: Mock is-my-json-valid validate schemas in tests

sirreal opened the pull request:
This PR mocks `is-my-json-valid` to include validation of the schemas received. Invalid schemas will throw, otherwise behavior is unchanged.

The result is that bad schemas are exposed in tests as long as the schema is passed to `is-my-json-valid` somewhere in the context of tests. That may well get most schema usage. The downside is that the source of the error is rather opaque, although the problem should be visible in the trace. For example:

```
 FAIL  client/state/data-layer/test/extensions-middleware.js
  ● Test suite failed to run

    TypeError: Invalid schema received

    data.properties[keys2.0] referenced schema does not match
    Found: {"id":"/properties/data","items":{"id":"/properties/data/items","properties":{"CTA":{"id":"/properties/data/items/properties/CTA","properties":{"hook":{"id":"/properties/data/items/properties/CTA/properties/hook","type":"string"},"message":{"id":"/properties/data/items/properties/CTA/properties/message","type":"string"},"primary":{"id":"/properties/data/items/properties/CTA/properties/primary","type":"boolean"}},"type":"object"},"content":{"id":"/properties/data/items/properties/content","properties":{"classes":{"id":"/properties/data/items/properties/content/properties/classes","type":"string"},"description":{"id":"/properties/data/items/properties/content/properties/description","type":"string"},"emblem":{"id":"/properties/data/items/properties/content/properties/emblem","type":"null"},"icon":{"id":"/properties/data/items/properties/content/properties/icon","type":"string"},"message":{"id":"/properties/data/items/properties/content/properties/message","type":"string"}},"type":"object"},"feature_class":{"id":"/properties/data/items/properties/feature_class","type":"string"},"id":{"id":"/properties/data/items/properties/id","type":"string"},"jitm_stats_url":{"id":"/properties/data/items/properties/jitm_stats_url","type":"string"},"template":{"id":"/properties/data/items/properties/template","type":"string"},"ttl":{"id":"/properties/data/items/properties/ttl","type":"integer"},"url":{"id":"/properties/data/items/properties/url","type":"string"}},"type":"object"},"type":"array"}
    Violates rule: {"$ref":"#"}

      35 | }
      36 |
    > 37 | validator.filter = ( schema, options ) => {
      38 |      throwOnInvalidSchema( schema );
      39 |      return imjv.filter( schema, options );
      40 | };


      at throwOnInvalidSchema (client/__mocks__/is-my-json-valid.js:37:170)
      at validator (client/__mocks__/is-my-json-valid.js:37:239)
      at makeParser (client/state/data-layer/wpcom-http/utils.js:72:46)
      at Object.<anonymous> (client/state/data-layer/wpcom/sites/jitm/index.js:199:33)
      at Object.<anonymous> (client/state/data-layer/wpcom/sites/index.js:15:13)
```

Alternative to #21486 
See also p4TIVU-8O2-p2

deBhal commented:
I really like this approach.

I've been using JSON Schema for front end validation (i.e. unrelated purposes), and this conversation has had me wondering if those schemas are valid - turns out they're not (my uris are invalid):
```
 FAIL  client/components/domains/registrant-extra-info/test/fr-validate-contact-details.js
  ● Test suite failed to run

    TypeError: Invalid schema received
    
    data.id must be uri format
    Found: "/contact-validation/fr"
    Violates rule: {"type":"string","format":"uri"}

      35 | }
      36 | 
    > 37 | validator.filter = ( schema, options ) => {
      38 | 	throwOnInvalidSchema( schema );
      39 | 	return imjv.filter( schema, options );
      40 | };
      
      
      at throwOnInvalidSchema (client/__mocks__/is-my-json-valid.js:37:170)
      at validator (client/__mocks__/is-my-json-valid.js:37:239)
      at Object.<anonymous> (client/components/domains/registrant-extra-info/fr-validate-contact-details.js:66:486)
      at Object.<anonymous> (client/components/domains/registrant-extra-info/test/fr-validate-contact-details.js:12:33)
```

deBhal commented on client/__mocks__/is-my-json-valid.js:
```diff
@@ -0,0 +1,40 @@
+import imjv from 'is-my-json-valid';
+import { forEach, get, isEmpty } from 'lodash';
+
+const validateSchema = imjv( jsonSchema, { verbose: true, greedy: true } );
```
It would be nice to squeeze "draft4" in here somewhere, maybe this variable name, in an attempt to save future devs that particular confusion.

sirreal replied:
👍  00d76c0

sirreal commented:
A pain point with this approach is clearly identifying which schema has a problem. I've attempted to fix this in 6276c5f by simply printing the whole schema. That seems to help.

sirreal commented on client/components/domains/registrant-extra-info/fr-schema.json:
```diff
@@ -1,5 +1,5 @@
 {
-	"id": "/contact-validation/fr",
+	"id": "https://public-api.wordpress.com/rest/v1/domains/validation-schema/fr",
```
This URL doesn't seem to resolve to anything. Is there any reason to keep the `id`, or can it just be removed?

[`id` in the spec](https://tools.ietf.org/html/draft-zyp-json-schema-04#section-7.2)

sirreal replied:
Ah-ha! #21524

deBhal replied:
My reading of the spec is that the id URI doesn't have to resolve to anything, but I intend to have the schema available there sooner or later, so it seems a good choice.

We do need to have some `id` to resolve `refs` when we're consuming on the backend. You can check out the `TLD_Validator` if you're curious.

sirreal replied:
Sounds good 👌 

Do you mind pulling these schema fixes into another PR? I'd like to keep this one focused.

deBhal replied:
No worries - https://github.com/Automattic/wp-calypso/pull/21588

deBhal replied:
That's deployed now, so you should be able to rebase it out of the history.

sirreal replied:
Thanks!

sirreal commented:
After cherry-picking #21420, it appears that all schemas "visible" to the changes in this PR are valid 😀

gwwar commented:
Oh clever, I didn't think to mock the entire library. I think this should cover most schemas, we use `is-my-json-valid` in `isValidStateWithSchema` and that's used by all default handlers. (Folks would need to explicitly provide custom handlers and improperly deserialize without checks to get around this.) 

I think your earlier unit test that tests serialize/deserialize for default state in the app should cover pretty much everything, with the main problem being as you mention tracking down which schema is misbehaving. I'll try giving it a spin.

gwwar commented:
Cool verified that running 👍  ` npm run test-client client/state/test/persistence.js` with a bad schema should throw. Printing the schema helps, we can probably add convention to add a title property for easier debugging. Can probably codemod the existing schemas with something generated from the filename.

sirreal commented:
Marked as `Needs review` as there have been votes in favor of this approach and I think it's ready to move forward.

Please ignore the schema fixes, those are being handled as independent PRs.

sirreal commented:
Great idea printing some brief identifying information @gwwar. I updated to print schema `title` and or `description` if provided, and fallback to printing the entire schema.

---

Rebased after all schema fixes have been merged. This PR now includes no schema fixes.
